File: JingRound/JingRoundView.m
//
//  JingRoundView.m
//  JingFM-RoundEffect
//
//  Created by isaced on 13-6-6.
//  Copyright (c) 2013年 isaced. All rights reserved.
//  By isaced:http://www.isaced.com/

#import "JingRoundView.h"
#import <QuartzCore/QuartzCore.h>
#import <CoreGraphics/CoreGraphics.h>

#define kRotationDuration 8.0

@interface JingRoundView ()

@property (strong, nonatomic) UIImageView *roundImageView;
@property (strong, nonatomic) UIImageView *playStateView;

@end

@implementation JingRoundView

- (id)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self initJingRound];
    }
    return self;
}

- (id)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];
    if (self) {
        [self initJingRound];
    }
    return self;
}

- (void) initJingRound {
    CGPoint center = CGPointMake(self.frame.size.width / 2.0, self.frame.size.height / 2.0);
    
    // set JingRoundView
    self.clipsToBounds = YES;
    self.userInteractionEnabled = YES;
    
    self.layer.cornerRadius = center.x;
    self.layer.borderWidth = 1.0;
    self.layer.borderColor = [[UIColor grayColor] CGColor];
    
    self.layer.shadowColor = UIColor.blackColor.CGColor;
    self.layer.shadowRadius = 2;
    self.layer.shadowOpacity = 0.6;
    self.layer.shadowOffset = CGSizeMake(0, 1);
    
    // set roundImageView
    UIImage *roundImage = self.roundImage;
    self.roundImageView = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, self.frame.size.width, self.frame.size.height)];
    [self.roundImageView setCenter:center];
    [self.roundImageView setImage:roundImage];
    [self addSubview:self.roundImageView];
    
    // set play state
    UIImage *stateImage;
    if (self.isRotating) {
        stateImage = [UIImage imageNamed:@"start"];
    }else{
        stateImage = [UIImage imageNamed:@"pause"];
    }
    
    self.playStateView = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, stateImage.size.width, stateImage.size.height)];
    [self.playStateView setCenter:center];
    [self.playStateView setImage:stateImage];
    [self addSubview:self.playStateView];
    
    // border
#if __IPHONE_OS_VERSION_MAX_ALLOWED > __IPHONE_6_1
#define kCGImageAlphaPremultipliedLast  (kCGBitmapByteOrderDefault | kCGImageAlphaPremultipliedLast)
#else
#define kCGImageAlphaPremultipliedLast  kCGImageAlphaPremultipliedLast
#endif
    CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceRGB();
    CGContextRef context = CGBitmapContextCreate(nil, self.frame.size.width, self.frame.size.height, 8, 0, colorSpace, kCGImageAlphaPremultipliedLast);
    CFRelease(colorSpace);
    
    CGContextSetStrokeColorWithColor(context, [[UIColor colorWithRed:1.0 green:1.0 blue:1.0 alpha:0.7] CGColor]);
    CGContextBeginPath(context);
    CGContextAddArc(context, center.x, center.y, center.x , 0, 2 * M_PI, 0);
    CGContextClosePath(context);
    CGContextSetLineWidth(context, 15.0);
    CGContextStrokePath(context);
    
    
    // convert the context into a CGImageRef
    CGImageRef image = CGBitmapContextCreateImage(context);
    CGContextRelease(context);
    UIImage* image2 = [UIImage imageWithCGImage:image];
    UIImageView *imgv =[[UIImageView alloc] initWithFrame:CGRectMake(0,0, self.frame.size.width , self.frame.size.height)];
    imgv.center = CGPointMake(self.frame.size.width/2, self.frame.size.height/2);
    imgv.image = image2;
    [self addSubview:imgv];
    
    //Rotation
    [self addRotationAnimation];
    
    _isRotating = YES;
}

- (void)addRotationAnimation {
    CABasicAnimation* rotationAnimation;
    rotationAnimation = [CABasicAnimation animationWithKeyPath:@"transform.rotation.z"];
    rotationAnimation.toValue = [NSNumber numberWithFloat: M_PI * 2.0];
    rotationAnimation.duration = kRotationDuration;
    rotationAnimation.repeatCount = FLT_MAX;
    rotationAnimation.cumulative = NO;
    rotationAnimation.removedOnCompletion = NO; // No Remove
    [self.roundImageView.layer addAnimation:rotationAnimation forKey:@"jinground.rotation"];
}

#pragma mark Setter

- (void)setRoundImage:(UIImage *)aRoundImage {
    _roundImage = aRoundImage;
    self.roundImageView.image = self.roundImage;
}

- (void)setRotationSpeed:(float)rotationSpeed {
    _rotationSpeed = rotationSpeed;

    self.roundImageView.layer.timeOffset = [self.roundImageView.layer convertTime:CACurrentMediaTime() fromLayer:nil];
    self.roundImageView.layer.beginTime = CACurrentMediaTime();
    self.roundImageView.layer.speed = rotationSpeed;
}

#pragma mark Action

- (void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event {
    if (_isRotating) {
        [self pause];
    }else{
        [self resume];
    }

    if ([self.delegate respondsToSelector:@selector(playStatuUpdate:)]) {
        [self.delegate playStatuUpdate:_isRotating];
    }
}

- (void)resume {
    CFTimeInterval pausedTime = [self.roundImageView.layer timeOffset];
    self.roundImageView.layer.speed = 1;
    self.roundImageView.layer.timeOffset = 0.0;
    self.roundImageView.layer.beginTime = 0.0;
    CFTimeInterval timeSincePause = [self.roundImageView.layer convertTime:CACurrentMediaTime() fromLayer:nil] - pausedTime;
    self.roundImageView.layer.beginTime = timeSincePause;
    
    [self setRotationSpeed:_rotationSpeed];
    
    _isRotating = YES;
    
    self.playStateView.image = [UIImage imageNamed:@"start"];
    [UIView animateWithDuration:0.25 animations:^{
        self.playStateView.alpha = 0.0;
    }];
}

- (void)pause {
    CFTimeInterval pausedTime = [self.roundImageView.layer convertTime:CACurrentMediaTime() fromLayer:nil];
    self.roundImageView.layer.speed = 0.0;
    self.roundImageView.layer.timeOffset = pausedTime;
    
    _isRotating = NO;
    
    self.playStateView.image = [UIImage imageNamed:@"pause"];
    [UIView animateWithDuration:0.25 animations:^{
        self.playStateView.alpha = 1.0;
    }];
}

@end

File: JingRound/JingRoundView.swift
//
//  JingRoundView.swift
//  JingFM-RoundEffect-Swift
//
//  Created by isaced on 14-6-11.
//  Copyright (c) 2014 Year isaced. All rights reserved.
//

import UIKit
import QuartzCore
import CoreGraphics

// Delegate
protocol JingRoundViewDelegate {
    func playStatuUpdate(playState:Bool)
}

// JingRoundView
class JingRoundView: UIView {
    
    var delegate:JingRoundViewDelegate?
    
    let rotationDuration:CFTimeInterval = 8.0
    
    var isPlay:Bool = false {
        didSet{
            if self.isPlay {
                startRotation()
            }else{
                pauseRotation()
            }
        }
    }

    var roundImage:UIImage!{
        didSet{
            roundImageView.image = roundImage
        }
    }
    
    var roundImageView:UIImageView!
    var playStateView:UIImageView!
    
    // init
    init(frame: CGRect) {
        super.init(frame: frame)
        initJingRound()
    }
    init(coder aDecoder: NSCoder!) {
        super.init(coder: aDecoder)
        
        initJingRound()
    }
    
    // init JingRound
    func initJingRound(){
        
        let center:CGPoint = CGPoint(x: self.frame.size.width / 2.0, y: self.frame.size.height / 2.0)
        
        //set JingRoundView
        self.clipsToBounds = true
        self.userInteractionEnabled = true
        
        self.layer.cornerRadius = center.x
        self.layer.borderColor = UIColor.lightGrayColor().CGColor
        self.layer.borderWidth = 1.0
        
        self.layer.shadowColor = UIColor.blackColor().CGColor
        self.layer.shadowRadius = 2;
        self.layer.shadowOpacity = 0.6;
        self.layer.shadowOffset = CGSizeMake(0, 1);
        
        //set roundImageView
        self.roundImageView = UIImageView(frame: CGRectMake(0, 0, self.frame.size.width, self.frame.size.height))
        self.roundImageView.center = center
        self.roundImageView.image = self.roundImage
        self.addSubview(self.roundImageView)
        
        //set play state
        var stateImage:UIImage;
        if self.isPlay {
            stateImage = UIImage(named: "start")
        }else{
            stateImage = UIImage(named: "pause")
        }
        
        self.playStateView = UIImageView(frame: CGRect(x: 0, y: 0, width: stateImage.size.width, height: stateImage.size.height))
        self.playStateView.center = center
        self.playStateView.image = stateImage
        self.addSubview(self.playStateView)

        //border
        var colorSpace:CGColorSpaceRef = CGColorSpaceCreateDeviceRGB();
        let bitmapInfo = CGBitmapInfo.fromRaw(CGImageAlphaInfo.PremultipliedFirst.toRaw())!
        var context:CGContextRef = CGBitmapContextCreate(nil, UInt(self.frame.size.width), UInt(self.frame.size.height), UInt(8), UInt(0), colorSpace, bitmapInfo)
        CFRelease(colorSpace);
        CGContextSetStrokeColorWithColor(context, UIColor.whiteColor().colorWithAlphaComponent(0.7).CGColor);
        CGContextBeginPath(context);
        CGContextAddArc(context, center.x, center.y, center.x, 0, CGFloat(2 * M_PI), 0)
        CGContextClosePath(context);
        CGContextSetLineWidth(context, 15.0);
        CGContextStrokePath(context);


        // convert the context into a CGImageRef
        var image:CGImageRef = CGBitmapContextCreateImage(context);
        CGContextRelease(context);
        var image2:UIImage = UIImage(CGImage: image)
        var imgv:UIImageView = UIImageView(frame: CGRect(x: 0, y: 0, width: self.frame.size.width, height: self.frame.size.height))
        imgv.center = CGPointMake(self.frame.size.width/2, self.frame.size.height/2);
        imgv.image = image2;
        self.addSubview(imgv)

        //Rotation
        var rotationAnimation = CABasicAnimation(keyPath: "transform.rotation.z")
        rotationAnimation.toValue = M_PI * 2.0
        rotationAnimation.duration = self.rotationDuration;
        rotationAnimation.repeatCount = HUGE
        rotationAnimation.cumulative = false;
        rotationAnimation.removedOnCompletion = false; //No Remove
        self.roundImageView.layer.addAnimation(rotationAnimation, forKey: "rotation")
        
        //pause
        if !self.isPlay {
            self.layer.speed = 0.0;
        }
        
    }
    
    func startRotation() {
        //start Animation
        self.layer.speed = 1.0;
        self.layer.beginTime = 0.0;
        
        let pausedTime:CFTimeInterval = self.layer.timeOffset
        let timeSincePause:CFTimeInterval = self.layer.convertTime(CACurrentMediaTime(), fromLayer: nil) - pausedTime
        
        self.layer.beginTime = timeSincePause;
        
        //set ImgView
        self.playStateView.image = UIImage(named: "start")
        UIView.animateKeyframesWithDuration(0.6, delay: 0, options: UIViewKeyframeAnimationOptions.CalculationModePaced, animations: {
                self.playStateView.alpha = 1;
            }, completion: { (finished:Bool) -> () in
                if finished {
                    UIView.animateWithDuration(1.0, animations: {
                        self.playStateView.alpha = 0
                        })
                }
            })
    }
    
    func pauseRotation() {
        //set ImgView
        self.playStateView.image = UIImage(named: "pause")
        self.userInteractionEnabled = false;
        
        UIView.animateWithDuration(0.6, animations: {
            self.playStateView.alpha = 1
        }, completion: { (finished:Bool) -> () in
            if finished {
                self.userInteractionEnabled = true;
                UIView.animateWithDuration(1.0, animations: {
                    self.playStateView.alpha = 0
                    //pause
                    let pausedTime:CFTimeInterval = self.layer.convertTime(CACurrentMediaTime(), fromLayer: nil)
                    self.layer.speed = 0.0;
                    self.layer.timeOffset = pausedTime;
                })
            }
        })
    }
    
    func play() {
        self.isPlay = true
    }
    
    func pause() {
        self.isPlay = false
    }
    
    override func touchesBegan(touches: NSSet!, withEvent event: UIEvent!) {
        self.isPlay = !self.isPlay;
        self.delegate?.playStatuUpdate(self.isPlay)
    }
}
